include/envoy/router/router.h

```diff
@@ -226,6 +229,36 @@ class HashPolicy {
                                           const Http::HeaderMap& headers) const PURE;
 };
 
+class MetadataMatchCriterion {
+public:
+  virtual ~MetadataMatchCriterion() {}
+
+  /*
+   * @return const std::string& the name of the metadata key
+   */
+  virtual const std::string& name() const PURE;
+
+  /*
+   * @return const Envoy::HashedValue& the value for the metadata key
+   */
+  virtual const HashedValue& value() const PURE;
+};
+
+typedef std::shared_ptr<const MetadataMatchCriterion> MetadataMatchCriterionConstSharedPtr;
```

Comment:
Does this need to be shared_ptr ?

Reply:
If a Route's RouteAction has `metadata_match` criteria and has multiple WeightedClusters with their own `metadata_match` criteria, the RouteAction's MetadataMatchCriterion ends being referenced by multiple MetadataMatchCriteria implementations.

That said, they could be copied instead of using a shared_ptr.

Reply:
That's fine, just wanted to check.